tensorflow/core/kernels/redux_functor.h

```diff
@@ -16,38 +16,208 @@ limitations under the License.
 #ifndef TENSORFLOW_CORE_KERNELS_REDUX_FUNCTOR_H_
 #define TENSORFLOW_CORE_KERNELS_REDUX_FUNCTOR_H_
 
-#include "third_party/eigen3/Eigen/Core"
 #include "tensorflow/core/framework/tensor_types.h"
 #include "tensorflow/core/platform/types.h"
+#include "third_party/eigen3/Eigen/Core"
 
 namespace tensorflow {
 
 using CPUDevice = Eigen::ThreadPoolDevice;
 
 namespace functor {
 
-// Compute reduction over all outer dimensions.
+// Compute reduction over outer dimensions.
 // Example:
-//   input: [32, 32, 256]
+//   input: [D1, D2, ... , DN]
 //   ->
-//   output: [256]
+//   output: [Di, ... , DN] where i belongs to set [1,N]
 template <typename T, typename AccumT, typename BinaryFunctor>
 struct ReduceOuterDimensions {
   template <int num_dims>
   void operator()(const CPUDevice& device,
                   const Eigen::DSizes<Eigen::Index, num_dims>& input_dims,
-                  const Tensor& input, Tensor* output) const {
-    static_assert(num_dims >= 2, "Input dimensions must at least 2");
-
+                  const Tensor& input, Tensor* output,
+                  const int num_reduce_dims = 1) const {
     // Compute inner and outer dim after reshaping into 2d tensor.
-    int64 inner_dim = input_dims[num_dims - 1];
-    int64 outer_dim = 1;
-    for (int i = 0; i < num_dims - 1; ++i) outer_dim *= input_dims[i];
+    int64 inner_dim = 1, outer_dim = 1;
+    for (int i = 0; i < num_dims - num_reduce_dims; ++i)
+      outer_dim *= input_dims[i];
+    for (int i = num_dims - num_reduce_dims; i < num_dims; ++i)
+      inner_dim *= input_dims[i];
+
+    if (1 == inner_dim) {
```

Comment:
Isn't it a bug? '32x32x1' should be reduced to scalar?

Reply:
Exactly! It should be "outer_dim" here.
Thank you very much for pointing out the bug!
Fixed.

Reply:
`tensorflow/python/keras:convolutional_test`

`ValueError: Shapes (3,) and (1, 1, 1, 1, 3) are incompatible`

Reply:
fixed